### PR TITLE
Update declarative assets for Nutanix

### PR DESCRIPTION
**Pull request by thunderboltsid**
- Update bootkube template to copy static resources for nutanix under data/data/bootstrap/files
- Add terraform files for Nutanix platform under data/data/nutanix
- Add terraform stages for Nutanix platform under pkg/terraform/stages/nutanix

### CODE REVIEW

File: data/data/nutanix/pre-bootstrap/main.tf [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  password     = var.password
+  endpoint     = var.prism_central
+  insecure     = var.insecure
+  port         = var.port
+}
+
+#TODO: Upload instead of query
+resource "nutanix_image" "rhcos" {
+  name        = var.nutanix_image
+  source_path  = var.nutanix_image_filepath
```
Comment (rvanderp3) [inline data/data/nutanix/pre-bootstrap/main.tf:17]: tf-fmt is failing
```suggestion
  source_path = var.nutanix_image_filepath
```

File: data/data/nutanix/master/main.tf [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  wait_timeout = 60
+  username     = var.username
+  password     = var.password
+  endpoint     = var.prism_central
+  insecure     = var.insecure
+  port         = var.port
+}
+
+resource "nutanix_virtual_machine" "vm_master" {
+  count = var.master_count
```
Comment (rvanderp3) [inline data/data/nutanix/master/main.tf:15]: tf-fmt is failing due '=' not lining up with the rest of the block
```suggestion
  count                = var.master_count
```

File: data/data/nutanix/variables-nutanix.tf [hunk 3]
```diff
@@ -0,0 +1,69 @@
+//////
+// Nutanix variables
+//////
+
+variable "prism_central" {
+  type        = string
+  description = "This is the Prism Central for the environment."
```
Comment (rvanderp3) [inline data/data/nutanix/variables-nutanix.tf:7]: Is this the hostname/IP for prism central?

Comment (rvanderp3) [reply data/data/nutanix/variables-nutanix.tf:7]: the description seems ambiguous

**Comment by rvanderp3** [comment]
> /lgtm
> 
> My only comment here is that it would have been helpful to have `tools.go` in its own commit.

apologies, this comment is not relevant to this PR

**Comment by rvanderp3** [comment]
@thunderboltsid would you mind taking a look at the merge conflict?

File: data/data/nutanix/bootstrap/main.tf [hunk 4]
```diff
@@ -0,0 +1,63 @@
+locals {
+  description = "Created By OpenShift Installer"
+}
+
+provider "nutanix" {
+  wait_timeout = 60
```
Comment (rvanderp3) [inline data/data/nutanix/bootstrap/main.tf:6]: Is a 60 minute creation timeout reasonable?  I would expect the bootstrap to be provisioned within a few minutes at most if all is well.  There may be a rationale behind this timeout that I'm not aware of.

File: data/data/nutanix/master/main.tf [hunk 5]
```diff
@@ -0,0 +1,50 @@
+locals {
+  description = "Created By OpenShift Installer"
+}
+
+provider "nutanix" {
+  wait_timeout = 60
```
Comment (rvanderp3) [inline data/data/nutanix/master/main.tf:6]: Is a 60 minute creation timeout reasonable?  I would expect the control plane nodes to be provisioned within a few minutes at most if all is well.  There may be a rationale behind this timeout that I'm not aware of.

File: data/data/nutanix/pre-bootstrap/main.tf [hunk 6]
```diff
@@ -0,0 +1,36 @@
+locals {
+  description = "Created By OpenShift Installer"
+}
+
+provider "nutanix" {
+  wait_timeout = 60
```
Comment (rvanderp3) [inline data/data/nutanix/pre-bootstrap/main.tf:6]: Is a 60 minute creation timeout reasonable?

File: data/data/nutanix/variables-nutanix.tf [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+variable "username" {
+  type        = string
+  description = "Prism Central user for the environment."
+}
+
+variable "password" {
+  type        = string
+  description = "Prism Central server password"
```
Comment (rvanderp3) [inline data/data/nutanix/variables-nutanix.tf:23]: Should this be `Prism Central user password`?

**Review by rvanderp3 — COMMENTED** [review]
Found a nit and had a few questions on the terraform provider timeout value.

File: data/data/nutanix/master/main.tf (hunk 5 above)
Comment (thunderboltsid) [reply data/data/nutanix/master/main.tf:6]: Good point. We haven't instrumented these stage times and as a result this is just a safe timeout. We can always reduce in future.

File: data/data/nutanix/bootstrap/main.tf (hunk 4 above)
Comment (thunderboltsid) [reply data/data/nutanix/bootstrap/main.tf:6]: Good point. We haven't instrumented these stage times and as a result this is just a safe timeout. We can always reduce in future.

File: data/data/nutanix/pre-bootstrap/outputs.tf [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  value = nutanix_image.bootstrap_ignition.id
+}
+
+output "cluster_domain" {
+  value = var.cluster_domain
+}
+
+output "cluster_id" {
+  value = var.cluster_id
+}
```
Comment (staebler) [inline data/data/nutanix/pre-bootstrap/outputs.tf:15]: Why are these outputs needed? They are not dependent on anything created by the stage.

File: data/data/nutanix/pre-bootstrap/main.tf [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  name        = var.nutanix_image
+  source_path = var.nutanix_image_filepath
+  description = local.description
+}
+
+resource "nutanix_category_key" "ocp_category_key" {
+  name = "openshift-${var.cluster_id}"
+}
+
+resource "nutanix_category_value" "ocp_category_value" {
```
Comment (staebler) [inline data/data/nutanix/pre-bootstrap/main.tf:23]: Should we add the `description` field to these?

File: data/data/nutanix/master/outputs.tf [hunk 10]
```diff
@@ -0,0 +1,3 @@
+output "control_plane_ips" {
+  value = nutanix_virtual_machine.vm_master.*.nic_list_status.0.ip_endpoint_list.0.ip
```
Comment (staebler) [inline data/data/nutanix/master/outputs.tf:2]: Let's use the new style for splat expressions.
```suggestion
  value = nutanix_virtual_machine.vm_master[*].nic_list_status[0].ip_endpoint_list[0].ip
```

File: pkg/terraform/stages/nutanix/stages.go [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+	"github.com/openshift/installer/pkg/terraform"
+	"github.com/openshift/installer/pkg/terraform/providers"
+	"github.com/openshift/installer/pkg/terraform/stages"
+)
+
+// PlatformStages are the stages to run to provision the infrastructure in Nutanix.
+var PlatformStages = []terraform.Stage{
+	stages.NewStage("nutanix", "pre-bootstrap", []providers.Provider{providers.Nutanix}),
+	stages.NewStage("nutanix", "bootstrap", []providers.Provider{providers.Nutanix}, stages.WithNormalBootstrapDestroy()),
+	stages.NewStage("nutanix", "master", []providers.Provider{providers.Nutanix}),
```
Comment (staebler) [inline pkg/terraform/stages/nutanix/stages.go:13]: Is there any technical reason why the "pre-bootstrap" and the "master" stages cannot be combined into a single stage which is run prior to the "bootstrap" stage? For example, will the nutanix terraform provider wait for the masters to fully ignite before the provider deems that master VMs to have been successfully created?

File: data/data/nutanix/variables-nutanix.tf [hunk 12]
```diff
@@ -0,0 +1,69 @@
+//////
+// Nutanix variables
+//////
+
+variable "prism_central_address" {
```
Comment (staebler) [inline data/data/nutanix/variables-nutanix.tf:5]: All of the variables in this file should have the "nutanix_" prefix.

**Review by staebler — COMMENTED** [review]
Please add OWNERS files at data/data/nutanix and pkg/terraform/stages/nutanix.

File: data/data/nutanix/pre-bootstrap/main.tf [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  name  = nutanix_category_key.ocp_category_key.id
+  value = "openshift-ipi-installations"
+}
+
+resource "nutanix_image" "bootstrap_ignition" {
```
[the remark this answers is not in the document]
Comment (thunderboltsid) [reply data/data/nutanix/pre-bootstrap/main.tf:28]: I don't think that's a good idea. Image upload to PC can be a rather slow process and sometimes fail due to network disruption. Case in which I'd rather have it fail early on.

File: data/data/nutanix/pre-bootstrap/main.tf [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+  name        = var.nutanix_bootstrap_ignition_image
+  source_path = var.nutanix_bootstrap_ignition_image
```
[the remark this answers is not in the document]
Comment (thunderboltsid) [reply data/data/nutanix/pre-bootstrap/main.tf:30]: No. Changing this to name and absolute path to make it clearer.

File: pkg/terraform/stages/nutanix/stages.go (hunk 11 above)
Comment (thunderboltsid) [reply pkg/terraform/stages/nutanix/stages.go:13]: The masters nodes get created after the bootstrap node gets created. The three stages are reflective of phases of the install pre-bootstrap(create images on Prism Central), bootstrap (provision bootstrap node), master (provision master nodes).

File: data/data/nutanix/pre-bootstrap/main.tf (hunk 13 above)
[the remark this answers is not in the document]
Comment (thunderboltsid) [reply data/data/nutanix/pre-bootstrap/main.tf:30]: Would it make sense to delete the image also as part of a successful cluster install workflow from the installer?

[the remark this answers is not in the document]
Comment (staebler) [reply data/data/nutanix/pre-bootstrap/main.tf:30]: Yes, the image should be deleted as part of a successful install as well as deleted as part of deleting a failed install.

File: pkg/terraform/stages/nutanix/stages.go (hunk 11 above)
Comment (staebler) [reply pkg/terraform/stages/nutanix/stages.go:13]: My question was whether there is any technical reason why the masters could not be created **before** the bootstrap node is created.

Comment (thunderboltsid) [reply pkg/terraform/stages/nutanix/stages.go:13]: None really, added a change to merge those two. See if you like it more this way.

File: data/data/nutanix/pre-bootstrap/outputs.tf (hunk 8 above)
Comment (thunderboltsid) [reply data/data/nutanix/pre-bootstrap/outputs.tf:15]: Cleaned up some outputs (were useful for debug output)

File: data/data/nutanix/pre-bootstrap/main.tf (hunk 13 above)
[the remark this answers is not in the document]
Comment (thunderboltsid) [reply data/data/nutanix/pre-bootstrap/main.tf:30]: Moved the ignition image to the bootstrap phase so now it gets deleted when bootstrap gets destroyed. Ignition image takes less than a minute to upload so shouldn't be a problem.

File: data/data/bootstrap/files/usr/local/bin/bootkube.sh.template [hunk 15]
```diff
@@ -337,6 +337,7 @@ then
 	copy_static_resources_for openstack
 	copy_static_resources_for ovirt
 	copy_static_resources_for vsphere
+	copy_static_resources_for nutanix
```
Comment (rvanderp3) [inline data/data/bootstrap/files/usr/local/bin/bootkube.sh.template:340]: can we pull this change in to it's own commit since its not related to the core terraform file addition?

Comment (staebler) [inline data/data/bootstrap/files/usr/local/bin/bootkube.sh.template:340]: This PR does not add any static Nutanix resources to copy. Let's hold off on adding the copy until there are files to copy.

File: data/data/nutanix/pre-bootstrap/main.tf [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    name  = nutanix_category_key.ocp_category_key.name
+    value = nutanix_category_value.ocp_category_value.value
+  }
+
+  guest_customization_cloud_init_user_data = base64encode(var.ignition_master)
+  nic_list {
+    subnet_uuid = var.nutanix_subnet_uuid
+  }
+
+  depends_on = [nutanix_image.rhcos]
```
Comment (staebler) [inline data/data/nutanix/pre-bootstrap/main.tf:73]: Is this depends_on needed? Do we not get this for free since the image is used in the disk list?

File: pkg/terraform/stages/nutanix/stages.go [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+
+import (
+	"github.com/openshift/installer/pkg/terraform"
+	"github.com/openshift/installer/pkg/terraform/providers"
+	"github.com/openshift/installer/pkg/terraform/stages"
+)
+
+// PlatformStages are the stages to run to provision the infrastructure in Nutanix.
+var PlatformStages = []terraform.Stage{
+	stages.NewStage("nutanix", "pre-bootstrap", []providers.Provider{providers.Nutanix}),
```
Comment (staebler) [inline pkg/terraform/stages/nutanix/stages.go:11]: Can we call this stage "cluster" instead of "pre-bootstrap"? The resources for the cluster are created by the stage.

File: data/data/nutanix/pre-bootstrap/main.tf [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+provider "nutanix" {
+  wait_timeout = 60
+  username     = var.nutanix_username
+  password     = var.nutanix_password
+  endpoint     = var.nutanix_prism_central_address
+  port         = var.nutanix_prism_central_port
+}
+
+resource "nutanix_category_key" "ocp_category_key" {
+  name        = "openshift-${var.cluster_id}"
```
Comment (staebler) [inline data/data/nutanix/pre-bootstrap/main.tf:14]: Is this similar to tags in AWS? If so, then we should stick with the common pattern of using "kubernetes.io/${var.cluster_id}/owned" as the key.

File: data/data/nutanix/pre-bootstrap/main.tf (hunk 16 above)
Comment (thunderboltsid) [reply data/data/nutanix/pre-bootstrap/main.tf:73]: Good catch; I don't think it is. TF should be able to infer this.

File: pkg/terraform/stages/nutanix/stages.go (hunk 17 above)
Comment (thunderboltsid) [reply pkg/terraform/stages/nutanix/stages.go:11]: Sure.

File: data/data/nutanix/pre-bootstrap/main.tf (hunk 18 above)
Comment (thunderboltsid) [reply data/data/nutanix/pre-bootstrap/main.tf:14]: That would not be a valid key format for Nutanix Prism Categories :)

File: data/data/bootstrap/files/usr/local/bin/bootkube.sh.template (hunk 15 above)
Comment (thunderboltsid) [reply data/data/bootstrap/files/usr/local/bin/bootkube.sh.template:340]: Yeah, seems like this is required to copy openshift resources. Files to copy are already in the codebase: i.e. coredns and keepalived templates.

File: data/data/nutanix/pre-bootstrap/main.tf (hunk 18 above)
Comment (staebler) [reply data/data/nutanix/pre-bootstrap/main.tf]: How about `kubernetes-io-cluster-${var.cluster_id}` for a key, like is used for gcp?

File: data/data/bootstrap/files/usr/local/bin/bootkube.sh.template (hunk 15 above)
Comment (staebler) [reply data/data/bootstrap/files/usr/local/bin/bootkube.sh.template:340]: Ah, I am mistaken about what this is copying. Sorry for the noise.

Comment (thunderboltsid) [reply data/data/bootstrap/files/usr/local/bin/bootkube.sh.template:340]: No worries. As discussed offline, we can keep the commit here.

File: data/data/nutanix/pre-bootstrap/main.tf (hunk 18 above)
Comment (thunderboltsid) [reply data/data/nutanix/pre-bootstrap/main.tf]: That works.

**Review by staebler — COMMENTED** [review]
/lgtm
/approve
